### PR TITLE
implemented and tested static stride scheduler for weighted round robin load balancing policy

### PR DESCRIPTION
**Pull request by tonyjongyoonan**
The core of the Weighted Round Robin load balancer policy on the client side is a stride scheduler originally implemented by an [EDFScheduler](http://google3/net/rpc2/edf_scheduler.h;rcl=438515966). However, the mutex lock required by the EDFScheduler has been a frequent source of thread contention at high request rates and a block on other cost saving efforts.

The Static Stride Scheduler is a generator of a practically equivalent sequence of picks as EDFScheduler. It removes the need for a priority queue (and thus a lock) and improves latency at high request rates. 

[go/static-stride-scheduler](https://g3doc.corp.google.com/production/rpc/stubs/g3doc/loadbalancing/static_stride_scheduler.md?cl=head)

### CODE REVIEW

**Comment by tonyjongyoonan** [comment]
notes: 
vectors not available

**Comment by ejona86** [comment]
You can continue pushing to your branch and update the PR as you go.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:469]: This should probably be an array, to avoid all the boxing of Long vs long. (I'm not trying to say whether this should be be 64 bit or 32 bit.)

Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:469]: We'll make these fields final.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:471]: This should be an AtomicInteger. If you haven't seen https://github.com/grpc/proposal/blob/master/A58-client-side-weighted-round-robin-lb-policy.md#earliest-deadline-first-edf-scheduler , there's a few things there. We don't have to do the _identical_ approach, but we'd avoid doing things differently when unnecessary.

`lambda () -> uint32 next_seq_fn` is "a function you call to get an int32. The equivalent of that in Java is `AtomicInteger.getAndIncrement()`. One important part of that function in the design is that it is shared across scheduler instances. If you re-create the scheduler for the same input data, you get the same results, so the code doesn't have to worry about poor weighting if it is re-created frequently. Since we didn't have that sort of state previously, the old scheduler had to randomize itself.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:504]: After the seq reaches Integer.MAX_VALUE , it rolls over and this becomes -1. The seq should be non negative.

The staticScheduler lives in the picker, and there might be multiple instances of picker in the system, so ideally the seq should be global among all the pickers. Alternatively, we need to randomize the weight during the static scheduler construction time.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:508]: rename to `pick()`

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:263]: volatile means the object is thread safe: if the object is accessed from different thread, JVM will make sure one operation at a time. This is needed due to `pick()` operation can be invoked from different threads.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:323]: I do the processing inside the scheduler constructor. I can definitely move the processing into the `updateWeight()` method, but the [documentation](https://g3doc.corp.google.com/production/rpc/stubs/g3doc/loadbalancing/static_stride_scheduler.md?cl=head#static-stride-scheduling) seems to suggest that we should be "building" the set of weights when we create the scheduler. Which would be correct?

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:276]: Should I remove the EDF scheduler entirely? Otherwise, I get an unused method style error.

**Comment by tonyjongyoonan** [comment]
TODO: I believe I need to fix some warnings (which might be failures) for parts of the edf scheduler that are now unused (ex. random). This probably requires changing the constructor as well as everywhere where the WRR LB is initialized.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:388]: Should be fixed.

Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:381]: Since you are storing 32 bits, instead of AtomicLong, use AtomicInteger. Then, you just need to cast _up_ to `long` as you read it out. That can be done with [`Integer.toUnsignedLong()`](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#toUnsignedLong-int-) or `seq & UINT32_MAX`. Masking instead of modulus makes it easier to avoid negative results. Then, instead of getAndUpdate(), just use getAndIncrement(). So this becomes:

```java
private final AtomicInteger sequence;

private long nextSequence() {
    return Integer.toUnsignedLong(sequence.getAndIncrement());
}
```

Note that I am paying attention to & UINT32_MAX being different from % UINT32_MAX. But we don't care about the difference here.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:288]: okay I see that you are processing it when you are building the weights array inside scheduler constructor. And that seems follow the [design](https://docs.google.com/document/d/1_YtzJzdOa-IDXnYRouzryKyp6sXmXGUKYayYZQ9U2KM/edit?resourcekey=0-OTzfjFCHe2IXHK-xA2Enbw):
> When less than two subchannels have load info, it degrades to round_robin.

That is fine. 
For the processing there is an optimization/alternative you can do to simplify the code and achieve the same thing, similar to the removed L301 - L305. You can count the non-zero weight channels instead and L362 becomes like:
```
if (weightedChannelCount >= 1) {
   avgWeight =  ... // scaled weight
} else {
   avgWeight = 1;
}
```

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:343]: Changed. As for your other comment, I was under the impression that AtomicInteger was signed, which would mean that it cannot support the max value for an unsigned 32 bit integer. Is this not the case?

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:381]: I was under the impression that AtomicInteger was signed, which would mean that it cannot support the max value for an unsigned 32 bit integer. Is this not the case?

Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:381]: It still has 32 bits and addition (the increment) is the same for signed and unsigned numbers. So the bits are what we'd hope they would be, but Java just won't treat them right in math. When casting to a larger integer size, then only difference between signed and unsigned is whether you sign extend: whether you fill the new bits on the left with 0s or 1s. For unsigned, you always use 0s. For signed, you copy the sign bit to the upper bits. So if you want unsigned conversion but only have signed, you do a signed conversion and then force the top bits to be 0s with the bitwise AND.

We could use AtomicLong and mask out the lower bits, but if we are masking the bits then AtomicInteger is just as good.

File: xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java:1091]: A test is needed when the sequence number wraps?

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:405]: this would potentially throw index out of bound, add verification in constructor.

**Review by YifeiZhuang — COMMENTED** [review]
A few more comments, it looks much better.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:405]: The value of the backendIndex is bounded by the sizeDivisor s.t. `backendIndex: [0,sizeDivisor)`, where `sizeDivisor == scaledWeights.length`. Thus, it is guaranteed to be within the bounds of the array. However, maybe it would be more appropriate to cast to an `int` earlier and make the backendIndex an `int` altogether? What do you think? This is of course under the assumption that we have less than 0xFFFF_FFFF backends.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:389]: I just noticed that this may generate negative value when the high bit is 1, e,g. `new AtomicInteger((int)2147483649L)` is `-2147483647`.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:405]: You are right, the size is an integer and the algorithm guarantees it will find an index in the array.
I think it is safe to cast it earlier to int, because it won't loose any precision. The max possible size is 0x7FFF_FFFFL (Integer.MAX_VALUE).

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:389]: So I haven't pushed the change yet because I'm still testing, but I changed this to `this.sequence = new AtomicInteger(this.random.nextInt());`. 

The previous code worked strangely due to the interaction of a double with a long as well as the int cast from a high value, and didn't achieve the desired behavior of generating a random number between 0 and `UINT32_MAX`.

The new code gives us values ranging from `Integer.MIN_VALUE` to `Integer.MAX_VALUE`. I believe this is the desired behavior, because when the sequence value is casted up in `nextSequence()`, negative values become positive values that become unsigned integers greater than `Integer.MAX_VALUE` due to the `Integer.toUnsignedLong()`.

Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:389]: It sounds fine. (You can just use `random` instead of `this.random`)

File: xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java:922]: typo

File: xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java:930]: I can understand that you wanted to test that when the `pick()` returns the index of the largest weight, you increment the counter. But I don't understand why you name the local variable `mean` - it looks like you are comparing the `pick()` result with the mean of the weight, but actually not. So if you rename the `mean` in this comparison to `largestWeightIndex`, it will make sense to me.

File: xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/WeightedRoundRobinLoadBalancerTest.java:1112]: In this case, you can reduce the boilerplate code by adding a constructor in FakeRandom and remove the other two  classes.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:354]: This value was chosen to have an array of shorts. I think we want that (`short[] scaledWeights`) since the precision of int really doesn't add much. Since Java doesn't have unsigned shorts, I think we should either make this value Short.MAX_VALUE (0x7FFF) or use `Short.toUnsignedInt()` when reading out of the array.

Since we only read and write to the array in one location, `Short.toUnsigned*()` makes a lot of sense to match the C impl and have the same amount of precision. If we interacted with the array frequently, using toUnsigned could be bug-prone.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:336]: This (L335-L337) is describing the priority queue implementation of EDF and is no longer the case here, but sounds like it is how current implementation is.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:379]: why this is necessary?

**Review by YifeiZhuang — APPROVED** [review]
LGTM.
Please get @ejona86  approval then merge.
Please consult @temawi to watch for interop test dashboard after merge.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:333]: I see why some of that stuff was removed, but the comment now has the same problem an earlier version of this code has: it name-drops EDF without actually saying how it is mapped to a scheduling problem. Let's just replace this line with "in which each object's deadline is the multiplicative inverse of the object's weight." I think that is the most important part of the mapping.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:400]: Annotate with `com.google.common.annotations.VisibleForTesting`

**Review by ejona86 — APPROVED** [review]

**Pull request merged**
